idc_index/index.py

```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+            "series_aws_url"
+        ].apply(self._extract_crdc_instance_uuid_from_series_aws_url)
+        # Perform left join on client.index using crdc_instance_uuid
+        merged_df = manifest_df.merge(
+            index_df_copy,
+            left_on="manifest_crdc_instance_uuid",
+            right_on="index_crdc_instance_uuid",
+            how="left",
+        )
+
+        # Create 'crdc_instance_uuid_found' column
+        merged_df["crdc_instance_uuid_found"] = merged_df[
+            "index_crdc_instance_uuid"
+        ].notna()
+
+        # Check if crdc_instance_uuid is found in the index
+        if not all(merged_df["crdc_instance_uuid_found"]):
+            missing_manifest_cp_cmds = merged_df.loc[
+                ~merged_df["crdc_instance_uuid_found"], "manifest_cp_cmd"
+            ]
+            missing_manifest_cp_cmds_str = f"The following manifest copy commands do not have any associated series in the index: {missing_manifest_cp_cmds.tolist()}"
+            raise ValueError(missing_manifest_cp_cmds_str)
+
+        # Create 'endpoint' column
+        merged_df["endpoint"] = merged_df.apply(
+            lambda row: aws_endpoint_url
+            if row["crdc_instance_uuid_found"]
+            and row["s3_url"] == row["series_aws_url"]
+            else gcp_endpoint_url,
+            axis=1,
+        )
+
+        # Check if there are more than one endpoints
+        if len(merged_df["endpoint"].unique()) > 1:
+            raise ValueError(
+                "The manifest contains URLs from both AWS and GCP. Please use only one provider."
+            )
+
+        # Calculate total size
+        total_size = merged_df["series_size_MB"].sum()
+        total_size = round(total_size, 2)
+
+        # Write a temporary manifest file
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as temp_manifest_file:
+            for s3_url in merged_df["s3_url"]:
+                temp_manifest_file.write(f"sync {s3_url} {downloadDir}\n")
+
+        return total_size, merged_df["endpoint"].iloc[0], Path(temp_manifest_file.name)
+
+    @staticmethod
+    def _track_download_progress(
+        size_MB: int, downloadDir: str, process: subprocess.Popen
+    ):
+        """
+        Track progress by continuously checking the downloaded file size and updating the progress bar.
+        """
+        total_size_bytes = size_MB * 10**6  # Convert MB to bytes
+
+        # Calculate the initial size of the directory
+        initial_size_bytes = sum(
+            f.stat().st_size for f in Path(downloadDir).iterdir() if f.is_file()
+        )
+
+        pbar = tqdm(
+            total=total_size_bytes,
+            unit="B",
+            unit_scale=True,
+            desc="Downloading data",
+        )
+
+        while True:
+            downloaded_bytes = (
+                sum(
+                    f.stat().st_size for f in Path(downloadDir).iterdir() if f.is_file()
+                )
+                - initial_size_bytes
```

Comment:
I guess it should be ok if some other process downloads/copies files to the same directory at the same time - the worst is it would jump to 100% and be stuck there - right?

Reply:
Download continues, progress bar stops and does not go over after reaching 100%. Resolving.